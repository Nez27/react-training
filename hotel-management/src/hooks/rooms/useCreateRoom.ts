import { useMutation, useQueryClient } from '@tanstack/react-query';
import toast from 'react-hot-toast';

// Services
import { createRoom as createRoomFn} from '@src/services/roomServices';

// Constants
import { ADD_SUCCESS } from '@src/constants/messages';

// Hooks
import { useUserRoomAvailable } from '../useUserRoomAvailable';

/**
 * Create room on database
 * @returns The boolean of isCreating and create room function
 */
const useCreateRoom = () => {
  const queryClient = useQueryClient();
  const { dispatch } = useUserRoomAvailable();

  const {
    mutate: createRoom,
    isPending: isCreating,
    isSuccess,
  } = useMutation({
    mutationFn: createRoomFn,
    onSuccess: (room) => {
      toast.success(ADD_SUCCESS);
      queryClient.invalidateQueries({ queryKey: ['rooms'] });

      // Add room available to global state
      dispatch!({
        type: 'addRoom',
        payload: [{ id: room.id, name: room.name }],
      });
    },
    onError: (err) => toast.error(err.message),
  });

  return {
    isCreating,
    createRoom,
    isSuccess
  };
};

export { useCreateRoom };
