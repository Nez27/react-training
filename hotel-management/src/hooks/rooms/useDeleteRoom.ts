import { useMutation, useQueryClient } from '@tanstack/react-query';
import toast from 'react-hot-toast';

// Services
import { deleteRoom as deleteRoomFn } from '@service/roomServices';

// Constants
import * as messages from '@constant/messages';

/**
 * Delete the room from database
 * @returns The boolean of status deleting and deleteRoom function
 */
const useDeleteRoom = () => {
  const queryClient = useQueryClient();

  const {
    isPending: isDeleting,
    mutate: deleteRoom,
    isSuccess,
  } = useMutation({
    mutationFn: deleteRoomFn,
    onSuccess: () => {
      toast.success(messages.DELETE_SUCCESS);
      queryClient.invalidateQueries({
        queryKey: ['rooms'],
      });
    },
    onError: (err) => toast.error(err.message),
  });

  return { isDeleting, deleteRoom, isSuccess };
};

export { useDeleteRoom };
