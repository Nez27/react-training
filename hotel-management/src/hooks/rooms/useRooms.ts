import toast from 'react-hot-toast';
import { useQuery, useQueryClient } from '@tanstack/react-query';
import { useSearchParams } from 'react-router-dom';

// Services
import { getAllRooms } from '@service/roomServices';

// Constants
import { DEFAULT_PAGE_SIZE } from '@constant/config';

/**
 * Fetch room from database
 * @returns The status of loading rooms from database and data of rooms
 */
const useRooms = () => {
  const queryClient = useQueryClient();
  const [searchParams] = useSearchParams();

  const sortByValue = searchParams.get('sortBy') || 'id';
  const orderByValue = searchParams.get('orderBy') || 'asc';
  const searchValue = searchParams.get('search') || '';
  const page = searchParams.get('page') ? Number(searchParams.get('page')) : 1;

  const {
    isLoading,
    data: { data: rooms, count } = {},
    error,
  } = useQuery({
    queryKey: ['rooms', sortByValue, orderByValue, searchValue, page],
    queryFn: () =>
      getAllRooms({
        sortBy: sortByValue,
        orderBy: orderByValue,
        roomSearch: searchValue,
        page,
      }),
  });

  if (error) {
    toast.error(error.message);
  }

  // Pre-fetching
  const totalPage = Math.ceil(count! / DEFAULT_PAGE_SIZE);

  if (page < totalPage) {
    const nextPage = page + 1;

    queryClient.prefetchQuery({
      queryKey: ['rooms', sortByValue, orderByValue, searchValue, nextPage],
      queryFn: () =>
        getAllRooms({
          sortBy: sortByValue,
          orderBy: orderByValue,
          roomSearch: searchValue,
          page: nextPage,
        }),
    });
  }

  if (page > 1) {
    const previousPage = page - 1;

    queryClient.prefetchQuery({
      queryKey: ['rooms', sortByValue, orderByValue, searchValue, previousPage],
      queryFn: () =>
        getAllRooms({
          sortBy: sortByValue,
          orderBy: orderByValue,
          roomSearch: searchValue,
          page: previousPage,
        }),
    });
  }

  return {
    isLoading,
    rooms,
    count
  };
};

export { useRooms };
