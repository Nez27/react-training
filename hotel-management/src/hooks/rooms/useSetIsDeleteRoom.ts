import { useMutation, useQueryClient } from '@tanstack/react-query';
import toast from 'react-hot-toast';

// Services
import { setIsDeleteRoom as setIsDeleteRoomFn } from '@service/roomServices';

// Constants
import * as messages from '@constant/messages';
import { useUserRoomAvailable } from '@hook/useUserRoomAvailable';

/**
 * Delete the room from database
 * @returns The boolean of status deleting and deleteRoom function
 */
const useSetIsDeleteRoom = () => {
  const queryClient = useQueryClient();
  const { dispatch } = useUserRoomAvailable();

  const {
    isPending: isDeleting,
    mutate: setIsDeleteRoom,
    isSuccess,
  } = useMutation({
    mutationFn: setIsDeleteRoomFn,
    onSuccess: (room) => {
      toast.success(messages.DELETE_SUCCESS);
      queryClient.invalidateQueries({
        queryKey: ['rooms'],
      });

      dispatch?.({ type: 'removeRoom', payload: [{ id: room.id }] });
    },
    onError: (err) => toast.error(err.message),
  });

  return { isDeleting, setIsDeleteRoom, isSuccess };
};

export { useSetIsDeleteRoom };
