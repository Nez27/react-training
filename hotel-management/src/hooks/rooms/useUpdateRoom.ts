import { useMutation, useQueryClient } from '@tanstack/react-query';
import toast from 'react-hot-toast';

// Services
import { updateRoom as updateRoomFn } from '@service/roomServices';

// Constants
import { UPDATE_SUCCESS } from '@constant/messages';
import { useUserRoomAvailable } from '@hook/useUserRoomAvailable';

/**
 * Update room from database
 * @returns The status of updating and updateRoom function
 */
const useUpdateRoom = () => {
  const queryClient = useQueryClient();
  const { dispatch } = useUserRoomAvailable();

  const {
    mutate: updateRoom,
    isPending: isUpdating,
    isSuccess
  } = useMutation({
    mutationFn: updateRoomFn,
    onSuccess: (room) => {
      toast.success(UPDATE_SUCCESS);
      queryClient.invalidateQueries({ queryKey: ['rooms'] });

      // Update name room in global state
      dispatch?.({
        type: 'updateRoomName',
        payload: [{ id: room.id, name: room.name }],
      });
    },
    onError: (err) => toast.error(err.message),
  });

  return { isUpdating, updateRoom, isSuccess };
};

export { useUpdateRoom };
