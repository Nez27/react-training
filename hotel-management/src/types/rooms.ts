interface IRoom {
  id: number;
  name: string;
  price: number;
  status: boolean;
}

export type { IRoom };
