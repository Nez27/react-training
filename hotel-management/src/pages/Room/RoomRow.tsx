import { useCallback, useMemo } from 'react';

// Helpers
import { formatCurrency } from '@helper/helper';

// Types
import { IRoom } from '@type/rooms';

// Components
import RoomForm from './RoomForm';
import Modal from '@component/Modal';
import Table from '@component/Table';
import Menus from '@component/Menus';
import { RiEditBoxFill } from 'react-icons/ri';
import { HiTrash } from 'react-icons/hi';
import ConfirmMessage from '@component/ConfirmMessage';

// Hooks
import { useDeleteRoom } from '@hook/rooms/useDeleteRoom';

// Constants
import { FORM } from '@constant/commons';

interface IRoomRow {
  room: IRoom;
}

const RoomRow = ({ room }: IRoomRow) => {
  const { id, name, price, status } = room;
  const { deleteRoom } = useDeleteRoom();
  const statusText = status 
    ? 'Unavailable' 
    : 'Available';
  const formattedPrice = useMemo(() => formatCurrency(price), [price]);
  const renderEditBtn = useCallback(
    (onCloseModal: () => void) => (
      <Menus.Button onClick={onCloseModal} icon={<RiEditBoxFill />}>
        Edit
      </Menus.Button>
    ),
    []
  );
  const renderDeleteBtn = useCallback(
    (onCloseModal: () => void) => (
      <Menus.Button icon={<HiTrash />} onClick={onCloseModal}>
        Delete
      </Menus.Button>
    ),
    []
  );

  return (
    <Table.Row>
      <div>{id}</div>
      <div>{name}</div>
      <div>{formattedPrice}</div>
      <div>{statusText}</div>

      <div>
        <Modal>
          <Menus.Menu>
            <Menus.Toggle id={id.toString()} />

            <Menus.List id={id.toString()}>
              <Modal.Open
                modalName={FORM.EDIT}
                renderChildren={renderEditBtn}
              />
              <Modal.Open
                modalName={FORM.DELETE}
                renderChildren={renderDeleteBtn}
              />
            </Menus.List>

            <Modal.Window name={FORM.EDIT} title="Edit Room">
              <RoomForm room={room} />
            </Modal.Window>

            <Modal.Window name={FORM.DELETE} title="Delete Room">
              <ConfirmMessage
                message={`Are you sure to delete ${name}?`}
                onConfirm={() => deleteRoom(id)}
              />
            </Modal.Window>
          </Menus.Menu>
        </Modal>
      </div>
    </Table.Row>
  );
};

export default RoomRow;
