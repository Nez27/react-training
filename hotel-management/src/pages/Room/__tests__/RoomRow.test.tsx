import renderer from 'react-test-renderer';
import RoomRow from '../RoomRow';
import { IRoom } from '@type/rooms';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

// Components

describe('RoomRow', () => {
  const tempRoom: IRoom = {
    id: 1,
    name: 'Temp Room',
    price: 250,
    status: true,
  };
  const queryClient = new QueryClient();

  const wrapper = renderer.create(
    <QueryClientProvider client={queryClient}>
      <RoomRow room={tempRoom} key={tempRoom.id} />
    </QueryClientProvider>
  );

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
