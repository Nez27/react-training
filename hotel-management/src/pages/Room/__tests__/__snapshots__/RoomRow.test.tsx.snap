// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomRow Should render correctly 1`] = `
<div
  className="sc-crzpnZ sc-bqGGcD gIMFtm jqrNRF"
>
  <div>
    1
  </div>
  <div>
    Temp Room
  </div>
  <div>
    $250.00
  </div>
  <div>
    Unavailable
  </div>
  <div>
    <div
      className="sc-fnVZQs CCxtf"
    >
      <button
        aria-label="Menu item 1"
        className="sc-bkbjWr fCztjK"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M10.5 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zm0 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zm0 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0z"
            fillRule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
